File: ScanbotSDKDemo/TIFFDemoViewController.m
//
//  TIFFDemoViewController.m
//  ScanbotSDKDemo
//
//  Created by Andrew Petrus on 08.02.18.
//  Copyright © 2018 doo GmbH. All rights reserved.
//

#import "TIFFDemoViewController.h"

@import ScanbotSDK;

@interface TIFFDemoViewController () <UINavigationControllerDelegate, UIImagePickerControllerDelegate>

@property (nonatomic, strong) NSMutableArray<UIImage *> *images;
@property (weak, nonatomic) IBOutlet UILabel *imagesCountLabel;

@end

@implementation TIFFDemoViewController

- (NSString *)applicationDocumentsDirectory{
    return [[[[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask] lastObject] path];
}

- (void)showResult:(BOOL)success filePath:(NSString *)filePath {
    UIAlertController *alert = nil;
    if (success) {
        alert = [UIAlertController alertControllerWithTitle:@"File saved"
                                                    message:[NSString stringWithFormat:@"TIFF file saved at path: %@", filePath]
                                             preferredStyle:UIAlertControllerStyleAlert];
    } else {
        alert = [UIAlertController alertControllerWithTitle:@"File not saved"
                                                    message:@"There was an error while saving TIFF file"
                                             preferredStyle:UIAlertControllerStyleAlert];
    }
    [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil]];
    [self presentViewController:alert animated:YES completion:nil];
}

- (BOOL)checkHasImages {
    if (self.images.count > 0) {
        return YES;
    }

    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Image(s) required"
                                                                   message:@"Please add at least one image from the Photo Library."
                                                            preferredStyle:UIAlertControllerStyleAlert];
    [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil]];
    [self presentViewController:alert animated:YES completion:nil];
    return NO;
}


- (IBAction)addImageButtoTapped:(id)sender {
    UIImagePickerController *picker = [[UIImagePickerController alloc] init];
    picker.delegate = self;
    [self presentViewController:picker animated:YES completion:nil];
}

- (IBAction)generateTIFFTapped:(id)sender {
    if (!self.checkHasImages) {
        return;
    }
    
    NSString *filePath = [NSString stringWithFormat:@"%@/%@.tiff",
                          [self applicationDocumentsDirectory], [[[NSUUID UUID] UUIDString] lowercaseString]];
    NSURL *fileURL = [NSURL URLWithString:filePath];
    
    SBSDKTIFFImageWriterParameters *params = [SBSDKTIFFImageWriterParameters defaultParameters];
    params.compression = COMPRESSION_LZW;
    params.dpi = 200;
    
    BOOL result = [SBSDKTIFFImageWriter writeTIFF:self.images fileURL:fileURL parameters:params];

    [self showResult:result filePath:filePath];
}

- (IBAction)generateBinarizedTIFFTapped:(id)sender {
    if (!self.checkHasImages) {
        return;
    }

    NSString *filePath = [NSString stringWithFormat:@"%@/%@_binarized.tiff",
                          [self applicationDocumentsDirectory], [[[NSUUID UUID] UUIDString] lowercaseString]];
    NSURL *fileURL = [NSURL URLWithString:filePath];
    
    SBSDKTIFFImageWriterParameters *params = [SBSDKTIFFImageWriterParameters defaultParameters];
    // alternatively init params via defaultParametersForBinaryImages, like:
    // SBSDKTIFFImageWriterParameters *params = [SBSDKTIFFImageWriterParameters defaultParametersForBinaryImages];
    params.binarize = YES;
    params.compression = COMPRESSION_CCITT_T6;
    params.dpi = 200;

    BOOL result = [SBSDKTIFFImageWriter writeTIFF:self.images fileURL:fileURL parameters:params];
    
    [self showResult:result filePath:filePath];
}

#pragma mark - UIImagePickerController delegate
- (void)imagePickerController:(UIImagePickerController *)picker
didFinishPickingMediaWithInfo:(NSDictionary<NSString *,id> *)info {
    UIImage *image = [info objectForKey:UIImagePickerControllerOriginalImage];
    if (!self.images) {
        self.images = [NSMutableArray array];
    }
    
    [self.images addObject:image];
    [self dismissViewControllerAnimated:YES
                             completion:^{
                                 self.imagesCountLabel.text = [NSString stringWithFormat:@"Images added: %lu", (unsigned long)self.images.count];
                             }];

}

@end

File: ScanbotSDKDemo/TIFFDemoViewController.swift
//
//  TIFFDemoViewController.swift
//  ScanbotSDKDemo
//
//  Created by Danil Voitenko on 25.11.21.
//  Copyright © 2021 doo GmbH. All rights reserved.
//

import UIKit
import ScanbotSDK

final class TIFFDemoViewController: UIViewController {
    @IBOutlet private var imagesCountLabel: UILabel?
    private var images: [UIImage] = []
    
    private func showFile(at filePath: String) {
        
    }
    
    @IBAction private func createTIFFDidPress(_ sender: Any) {
        createTIFF(isBinarized: false)
    }
    
    @IBAction private func createBinarizedTIFFDidPress(_ sender: Any) {
        createTIFF(isBinarized: true)
    }
    
    @IBAction private func importImageButtonDidPress(_ sender: Any) {
        let imagePicker = UIImagePickerController()
        imagePicker.sourceType = .photoLibrary
        imagePicker.delegate = self
        imagePicker.allowsEditing = false
        present(imagePicker, animated: true)
    }
    
    private func createTIFF(isBinarized: Bool) {
        guard let documentsDirectory = FileManager.default.urls(for: .documentDirectory,
                                                                   in: .userDomainMask).last?.path else { return }
        let filePath = "\(documentsDirectory)/\(UUID().uuidString.lowercased()).tiff"
        guard let fileURL = URL(string: filePath) else { return }
        let parameters = SBSDKTIFFImageWriterParameters.default()
        parameters.dpi = 200
        if isBinarized {
            parameters.binarize = true
            parameters.compression = SBSDKTIFFImageWriterCompressionOptions.COMPRESSION_CCITT_T6
            parameters.userDefinedFields = [SBSDKTIFFImageWriterUserDefinedField(stringValue: "String value",
                                                                                 fieldName: "SomeStringField",
                                                                                 fieldTag: 65000),
                                            SBSDKTIFFImageWriterUserDefinedField(numericValue: NSNumber(value: 123),
                                                                                 fieldName: "SomeIntField",
                                                                                 fieldTag: 65001),
                                            SBSDKTIFFImageWriterUserDefinedField(numericValue: NSNumber(value: 123.5),
                                                                                 fieldName: "SomeDoubleField",
                                                                                 fieldTag: 65535)]
        } else {
            parameters.compression = SBSDKTIFFImageWriterCompressionOptions.COMPRESSION_LZW
        }
        if SBSDKTIFFImageWriter.writeTIFF(images, fileURL: fileURL, parameters: parameters) {
            let alert = UIAlertController(title: "File saved",
                                          message: "At path: \(fileURL.path)",
                                          preferredStyle: .alert)
            let okAction = UIAlertAction(title: "OK",
                                         style: .default,
                                         handler: nil)
            alert.addAction(okAction)
            present(alert, animated: true, completion: nil)
        }
    }
}

extension TIFFDemoViewController: UIImagePickerControllerDelegate, UINavigationControllerDelegate {
    func imagePickerController(_ picker: UIImagePickerController,
                               didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
        if let image = info[UIImagePickerController.InfoKey.originalImage] as? UIImage {
            images.append(image)
        }
        picker.dismiss(animated: true) { [weak self] in
            self?.imagesCountLabel?.text = "Images added: \(self?.images.count ?? 0)"
        }
    }
}
